### PR TITLE
Added functionnalities to mid node (fabric-client 1.4.0)

### PR DESCRIPTION
**Pull request by Andromelus**
# fabric-config node
Identity name and display name of the configuration are separated, allowing the user to insert a custom configuration name. This is useful in case the user uses identities from two organizations but with the same name (org1 admin and org2 admin for example). Previously, we could only see two "admin" configurations, forcing us to open the configuration to check which config correspond to which org.

# fabric-mid node

## Dynamic configuration
This node now handles dynamic configuration. It can be configured via the `msg.payload` object. 
Any configuration parameter passed to the msg.payload object will overwrite any manual configuration of the node.
For the full list of the configuration parameters, depending on what action the node must perform, please check node description

## Submit and Evaluate transaction
No change. They still use `fabric-network 1.4.0`

## Listen for event(s)

It uses `fabric-client 1.4.0` 

This node can listen for events. There are two "modes" of listening. This mode can be set with the timeout configuration parameter.

The default mode is `timeout = true;`

### Listen for ever

`timeout = false`

If timeout is false, the listener never disconnect/unregister and keep listening. All the events are sent one by one to the next node in `msg.payload`.

**Warning** : If you specified an interval with an end block, the listener **WILL NOT** disconect or unregister. If an interval with an end block is specified, it is preferable to set `timeout = true`.

### Listen "not for ever"

This is the default mode, with `timeout = true`.

The listener reset every time a transaction is found. The timeout's value is 2 seconds.

The timeout is needed because of a bug with the listener and because of the way it works.
More information here: https://jira.hyperledger.org/browse/FABN-1207

The events are sent to the next node all together in an array in `msg.payload`.

## Query world state

The mid node can now query the world state. This uses `fabric-client 1.4.0`.

## Query block

The mid node can now query a block in the ledger. Block is queried by block number.
This uses `fabric-client 1.4.0`.

## Query transaction

The mid node can now query a transaction in the ledger. Transaction is queried by transaction ID.
This uses `fabric-client 1.4.0`.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -1,6 +1,6 @@
 {
     "name": "node-red-contrib-fabric",
-    "version": "0.0.1",
+    "version": "0.0.3",
```
Comment (cazfletch) [inline package.json]: why 0.0.3? shouldn't this be 0.0.2

File: nodes/fabric.js [hunk 2]
```diff
@@ -74,12 +74,35 @@ module.exports = function (RED) {
...
+    }
+
+    /**
+     * @param {object} channel the channel object
+     * @param {object} payload payload
+     * @param {string} contractName the name of the contrat
+     * @param {Node} node node
+     * @returns {Promise<Buffer>} promise
+     */
+    function query(channel, payload, contractName, node) {
```
Comment (cazfletch) [inline nodes/fabric.js]: you don't need this it does the same as evaluate

File: nodes/fabric.js [hunk 3]
```diff
@@ -121,32 +147,166 @@ module.exports = function (RED) {
...
+     * It also can listen on an interval of blocks, useful in cases you do not want to keep listenning
+     * Sends all the events in an array or one by one in the msg.payload depending on the configuration
+     * @param {peer} peer The peer to connect to
+     * @param {channel} channel The channel on which create the event hub
+     * @param {string} chaincodeName The name of the chaincode
+     * @param {Node} node node
+     * @param {object} msg the msg object
+     * @returns {Promise<Buffer>} promise
+     */
+    async function subscribeToEvent(peer, channel, chaincodeName, node, msg) {
```
Comment (cazfletch) [inline nodes/fabric.js]: I think you can combine the two functions. They both do the same thing

File: nodes/fabric.js [hunk 4]
```diff
@@ -121,32 +147,166 @@ module.exports = function (RED) {
...
+     *
+     * @param {string} identityName identityName
+     * @param {string} channelName channel
+     * @param {string} orgName name of the organization
+     * @param {string} peerName name of the peer
+     * @param {object} connection the connection config
+     * @param {string} walletLocation the wallet location (the files, not the folder)
+     * @returns {PromiseLike<Contract | never>} promise
+     */
+    async function connectToPeer(identityName, channelName,
```
Comment (cazfletch) [inline nodes/fabric.js]: you don't need this function. You can get the channel from the network. You don't need the peer as you don't need to create a new event hub you can just get them like the subscribeToEvents function does, and you already have the identity from the wallet

**Review by cazfletch — CHANGES_REQUESTED** [review]
Thanks for the PR. You have some duplicate functionality and also are making it more complicated than it needs to be

File: nodes/fabric.js (hunk 3 above)
Comment (Andromelus) [reply nodes/fabric.js]: That is what I wanted to do first, but they work in a slightly different context. The original one (yours) starts right when the flow starts (in node). However, the version I made starts only when it receives a message from the precedent node. I've not checked that in details yet. I'll come back to you if I have issues with it.

File: nodes/fabric.js (hunk 2 above)
Comment (Andromelus) [reply nodes/fabric.js]: True. I did not pay attention to that. I'll remove the query.

File: package.json (hunk 1 above)
Comment (Andromelus) [reply package.json]: Mistake from me.

File: nodes/fabric.js (hunk 4 above)
Comment (Andromelus) [reply nodes/fabric.js]: Correct. I'll remove it and use yours.

File: nodes/fabric.html [hunk 5]
```diff
@@ -50,140 +48,262 @@
...
-        <input type="text" id="node-input-contractName" placeholder=""/>
+        <input type="text" id="node-input-contractName" placeholder="" />
     </div>
-
     <div class="form-row">
         <label for="node-input-eventName"><i class="fa"></i>Event Name</label>
-        <input type="text" id="node-input-eventName" placeholder=".*"/>
+        <input type="text" id="node-input-eventName" placeholder=".*" />
+    </div>
+    <div class="form-row">
```
Comment (cazfletch) [inline nodes/fabric.html]: can you make it obvious that these properties are optional see https://nodered.org/docs/creating-nodes/help-style-guide. You could also maybe make them in a twisty?

**Review by cazfletch — CHANGES_REQUESTED** [review]
Looks good apart from the optional settings on the input node

File: nodes/fabric.html [hunk 6]
```diff
@@ -238,23 +239,54 @@
...
-        </ul>
-        </li>
-        <li>Target peer
-            <ul>
-                <li> If the <code>peerName</code> value is <b>NOT</b> set, the node will listen on all known peers (defined in your connection profile). As a consequence, you will most likely have duplicates events. Note that the node does not indicates from
-                    which peer an event comes from.</li>
-                <li> If the <code>peerName</code> value is set, the node will listen only on the peer with the given name.</li>
-            </ul>
-        </li>
+    <p>Hyperledger Fabric input node. Subscribe to events emitted by a blockchain</p>
```
Comment (cazfletch) [inline nodes/fabric.html:246]: I'm not sure its obvious enough that these are optional, there is nothing on the node so you have to read the info

**Review by cazfletch — CHANGES_REQUESTED** [review]

**Pull request merged**